File: population.py
from __future__ import division
from dna import DNA
import string
import random


""" 




"""
class Population() :
    
    
    def __init__(self) :
        
        self.population    = []       # the list contain the population
        self.matingPool    = []       # the list that will conatin the mating          
        self.generations   = 0        # the number of generations    
        self.finished      = False    # to check when we are done         
        self.perfectScore  = 0
        self.increments = 0.005
        
    #1 - initilize, the elements of the population are the dna  
    def population_init(self, target, mutationRate, population_size, init_score= 0.09)    :
        
        self.init_score= init_score
        self.target = target
        self.mutationRate = mutationRate
        self.population_size = population_size
        self.population = [""]* self.population_size
        
        for i in range(population_size) :
            self.population[i] = DNA(dna_size=len(self.target))
            self.population[i].dna()
        
        self.calcFitness()    
        self.matingPool    = [] 
        self.generations = 0
        self.finished = False
        self.perfectScore = 1
           
    #2- calculate the fitness
    def calcFitness(self) :
        
        for i in range(len(self.population)) : 
            self.population[i].fitness(self.target)
            
            
    #3- generate the mating pool
    def naturalSelection(self) :
        self.matingPool    = [] 
        
        #3.1 get the maxfitness
        self.maxFitness = 0.0
        for i in range(len(self.population)) :
            if self.population[i].fitness_score > self.maxFitness :
                self.maxFitness = self.population[i].fitness_score
                
    
        #3.1 // Based on fitness, each member will get added to the mating pool a certain number of times
        #    // a higher fitness = more entries to mating pool = more likely to be picked as a parent
        #    // a lower fitness = fewer entries to mating pool = less likely to be picked as a parent
        
        for i in range(len(self.population)) :
            # map fitness between 0 and 1
            
            fitness = self.population[i].fitness_score/self.maxFitness
            
            n = int(fitness * 100)
            for i in range(n) :
                if self.population[i].fitness_score  >= self.init_score :
                    
                    self.matingPool.append(self.population[i])
        
        if not len(self.matingPool) :
            self.init_score = self.init_score - self.increments*2.5
            self.naturalSelection()
    #4- create the new generation
    def generate(self, target)  :
        
        # 4.1 // Refill the population with children from the mating pool
        for i in range(len(self.population)):
         
         
            a = random.randint(0, len(self.matingPool) -1)
            b = random.randint(0, len(self.matingPool) -1)

            partnerA = self.matingPool[a]
            partnerB = self.matingPool[b]
            
         
            #4.2 crossover
            child = partnerA.crossover(partnerB)
            #4.3 mutate
            child.mutation(self.mutationRate, target)
            self.population[i] = child

        self.generations += 1
        
        if self.generations%1==0 :
            self.init_score += self.increments*1.5
        print("score :", self.init_score)
    def getBest(self) -> string:
        
        worldrecord = 0 
        index = 0
        for i in range(len(self.population)) :
            if self.population[i].fitness_score > worldrecord :
                worldrecord = self.population[i].fitness_score
                index = i
                #print("score :", self.population[index].fitness_score, self.population[index].getPhrase())
                
        if worldrecord == self.perfectScore :
            self.finished = True
 
        return self.population[index].getPhrase()

File: Population.cpp
#include <iostream>
#include "Population.hpp"
#include <random>


Population::Population()
{

}


Population::~Population()
{

   population.clear();
   matingPool.clear() ;
   std::cout << "END\n" ;
}

Population::Population(std::string target_t, float mutationRate_m, int population_p, float threshold_score_ts){

    
    set_POP_MAX(population_p);
    set_Target(target_t);
    set_MUTATION_RATE(mutationRate_m);
    set_THRESHOLD_SCORE(threshold_score_ts) ;
    for (int i = 0 ; i < population_max; i++){
        population.push_back(new DNA(target_t.length()));
        population[i]->fitness(target_phrase);

    }
    
    matingPool = std::vector <DNA*>();
    perfectScore = 1 ;
    finished = false;
    generations = 0 ;
} 


 void Population::calcFitness(){

    for (int i = 0 ; i < population_max; i++){
        population[i]->fitness(target_phrase);

    }

 }


void Population::naturalSelection() {
    matingPool.clear();
    float maxFitness = 0.0;
    for (int i = 0; i < population.size(); i++) {
      if (population[i]->fitness_score > maxFitness) {
        maxFitness = population[i]->fitness_score;
      }
    }
    
    // Based on fitness, each member will get added to the mating pool a certain number of times
    // a higher fitness = more entries to mating pool = more likely to be picked as a parent
    // a lower fitness = fewer entries to mating pool = less likely to be picked as a parent
    for (int i = 0; i < population.size(); i++) {
      
       
      float fitness = population[i]->fitness_score/maxFitness ;
      int n = int(fitness * 100);  // Arbitrary multiplier, we can also use monte carlo method
      for (int j = 0; j < n; j++) {              // and pick two random numbers
        if (population[i]->fitness_score > threshold_score) {

            matingPool.push_back(population[i]);
            std::cout << "i :"<< i << " -->" <<population[i]->fitness_score<<std::endl;

        }
        
      }
    }


    if (matingPool.size() == 0) {
        threshold_score = threshold_score - increments*2.5 ;
        naturalSelection() ;

    }
    //std::cout << "D:"<<maxFitness<<std::endl;
    std::cout << "B:"<<matingPool.size()<< " -->" <<threshold_score<<std::endl;
    for (int i = 0; i < matingPool.size(); i++) {
        std::cout << "C:"<<matingPool[i]->GENES() <<std::endl;
        
    }

}

void Population::generate(){
    DNA* child ;
    for (int i = 0; i < population.size(); i++){
        int m = matingPool.size() -1 ;
        int a = random(m);
        int b = random(m);
        std::cout << a << " <--> " << b << " <-->" << m << std::endl;
        DNA* partnerA = matingPool[a];
        DNA* partnerB = matingPool[b] ;

        std::cout << partnerA->GENES() << " == " << partnerB->GENES() << " ==>" << m << std::endl;
        //child = partnerA->crossover(*partnerB) ;
        //child->mutate(mutationRate);
        //population[i] = child;
    }

    generations++ ;
    threshold_score += increments ;


}


 std::string* Population::getBest(){

    int worldrecord = 0 ;
    int index = 0 ;
    for (int i = 0; i < population.size(); i++){
        if(population[i]->fitness_score > worldrecord){
            worldrecord = population[i]->fitness_score;
            index = i;

        }

    }
    if(worldrecord==perfectScore){
        finished = true;

    }


    return population[index]->getPhrase() ;
 }

void Population::set_THRESHOLD_SCORE(float score){
    threshold_score = score;
}
float Population::get_THRESHOLD_SCORE() {

    return threshold_score;
}



void Population::set_POP_MAX(int num){
    population_max = num;

}

int Population::get_POP_MAX() {
    return population_max;

}


void Population::set_Target(std::string phrase){
    target_phrase = phrase;

}
std::string Population::get_Target() {

    return target_phrase;
}




void Population::set_MUTATION_RATE(float mutation_rate){
    mutationRate = mutation_rate ;
}
float Population::get_MUTATION_RATE() {

    return mutationRate ;
}

char Population::random(int min, int max)
{
    std::random_device rd; // obtain a random number from hardware
    std::mt19937 generator(rd()); // seed the generator
    std::uniform_int_distribution<> distr(96, 110); // define the range

    return (char) distr(generator);
}


int Population::random(int max)
{
    std::random_device rd; // obtain a random number from hardware
    std::mt19937 generator(rd()); // seed the generator
    std::uniform_int_distribution<> distr(0, max); // define the range

    return distr(generator);
}

float Population::random()
{
    std::random_device rd; // obtain a random number from hardware
    std::mt19937 generator(rd()); // seed the generator
    std::uniform_real_distribution<> distr(0, 1); // define the range

    return (float) distr(generator);
}
